staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go

```diff
@@ -622,3 +531,163 @@ func TestInvalidKeys(t *testing.T) {
+	}{
+		{
+			name:              "list 10 corev1.Pod",
+			objectNum:         10,
+			newObjectFunc:     generatePod,
+			newListObjectFunc: func() runtime.Object { return &corev1.PodList{} },
+		},
+		{
+			name:              "list 100 corev1.Pod",
+			objectNum:         100,
+			newObjectFunc:     generatePod,
+			newListObjectFunc: func() runtime.Object { return &corev1.PodList{} },
+		},
+		{
+			name:              "list 1000 corev1.Pod",
+			objectNum:         1000,
+			newObjectFunc:     generatePod,
+			newListObjectFunc: func() runtime.Object { return &corev1.PodList{} },
+		},
+		{
+			name:              "list 10000 corev1.Pod",
+			objectNum:         10000,
+			newObjectFunc:     generatePod,
+			newListObjectFunc: func() runtime.Object { return &corev1.PodList{} },
+		},
+		{
+			name:              "list 50000 corev1.Pod",
+			objectNum:         50000,
+			newObjectFunc:     generatePod,
+			newListObjectFunc: func() runtime.Object { return &corev1.PodList{} },
+		},
+		{
+			name:              "list 10 corev1.Node",
+			objectNum:         10,
+			newObjectFunc:     generateNode,
+			newListObjectFunc: func() runtime.Object { return &corev1.NodeList{} },
+		},
+		{
+			name:              "list 100 corev1.Node",
+			objectNum:         100,
+			newObjectFunc:     generateNode,
+			newListObjectFunc: func() runtime.Object { return &corev1.NodeList{} },
+		},
+		{
+			name:              "list 1000 corev1.Node",
+			objectNum:         1000,
+			newObjectFunc:     generateNode,
+			newListObjectFunc: func() runtime.Object { return &corev1.NodeList{} },
+		},
+		{
+			name:              "list 10000 corev1.Node",
+			objectNum:         10000,
+			newObjectFunc:     generateNode,
+			newListObjectFunc: func() runtime.Object { return &corev1.NodeList{} },
+		},
+		{
+			name:              "list 50000 corev1.Node",
+			objectNum:         50000,
+			newObjectFunc:     generateNode,
+			newListObjectFunc: func() runtime.Object { return &corev1.NodeList{} },
+		},
+	}
+	for _, tc := range testCases {
+		b.Run(tc.name, func(b *testing.B) {
+			// booting etcd instance
+			ctx, store, etcdClient := testSetup(b)
+			defer etcdClient.Close()
+
+			// make fake objects..
+			dir := "/testing"
+			key := dir + "/testkey"
+			originalRevision := ""
+			for i := 0; i < tc.objectNum; i++ {
+				obj := tc.newObjectFunc()
+				o := obj.(metav1.Object)
+				o.SetName(fmt.Sprintf("object-%d", i))
+				o.SetSelfLink("testlink")
+				key := fmt.Sprintf("%s/%s", key, o.GetName())
+				out := tc.newObjectFunc()
+				if err := store.Create(ctx, key, obj, out, 0); err != nil {
+					b.Fatalf("Set failed: %v", err)
+				}
+				originalRevision = out.(metav1.Object).GetResourceVersion()
+			}
+			list := tc.newListObjectFunc()
+
+			// now we start benchmarking
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				if err := store.GetList(ctx, dir, storage.ListOptions{Predicate: storage.Everything, Recursive: true}, list); err != nil {
```

Comment:
This is still returning all objects.
I wanted to see a case where we return only a relatively small subset of objects (say 1% or sth like that).

Reply:
fixed, PTAL @wojtek-t